### PR TITLE
Generate custom tags for oneof fileds (#611)

### PR DESCRIPTION
**Pull request by krhubert**

### CODE REVIEW

File: protoc-gen-gogo/generator/generator.go [hunk 1]
```diff
@@ -3366,6 +3364,9 @@ func (g *Generator) generateMessage(message *Descriptor) {
 
 			oneofField := oFields[*field.OneofIndex]
 			tag = "protobuf:" + g.goTag(message, field, wiretype)
+			if gogoMoreTags != nil {
+				tag = " " + *gogoMoreTags
```
Comment (jmarais) [inline protoc-gen-gogo/generator/generator.go]: I think we needed a ```+=``` here.

**Comment by jmarais** [comment]
Hi. Thanks for the pullrequest. It seems like the oneof tags were collateral damage of a large generator refactor.

File: protoc-gen-gogo/generator/generator.go [hunk 2]
```diff
@@ -3366,6 +3364,9 @@ func (g *Generator) generateMessage(message *Descriptor) {
 
 			oneofField := oFields[*field.OneofIndex]
 			tag = "protobuf:" + g.goTag(message, field, wiretype)
+			if gogoMoreTags != nil {
```
Comment (jmarais) [inline protoc-gen-gogo/generator/generator.go]: I think we can just remove any tag logic here. 
I see the tags for every message does get constructed here:
https://github.com/gogo/protobuf/blob/23325ce401c6d7afd7008f26ff87c20074ea1413/protoc-gen-gogo/generator/generator.go#L2838-L2852

We just need to use that ```tag``` for our oneofSubField struct.

**Comment by jmarais** [comment]
Can we please also get a test for this issue?

I think it would make sense to add a simple oneof message under"
```test/tags/tags.proto```
And then use a xml custom tag with a marshal/unmarshal like the other tests to validate that the correct ```customtag``` has been generated.

**Comment by krhubert** [comment]
@jmarais I think it's ready to review.

In tests I skip the Marshal for json and XML because I will have to implement custom marshaler for oneof interface.

**Comment by krhubert** [comment]
About comments for tests look on the answer below.

> Do you have a suggestion for a test in order to validate a custom moretags field?

What do you think about using reflection and get a tag, instead of marshal/unmarshal with json/xml?

> With regards to the travis job. Remember to make in the root of the project in order to regenerate all the protos with the generator change.

Yes, I miss that. Also I found something interesting - in tags.proto syntax is set to [proto2](https://github.com/gogo/protobuf/blob/master/test/tags/tags.proto#L29), but the generated `test.pb.go` declares [GoGoProtoPackageIsVersion3](https://github.com/gogo/protobuf/blob/master/test/tags/tags.pb.go#L22) - why it so?

**Comment by jmarais** [comment]
> What do you think about using reflection and get a tag, instead of marshal/unmarshal with json/xml?
That is a good suggestion. I think that will make for a nice test.

> but the generated test.pb.go declares GoGoProtoPackageIsVersion3 - why it so?
That is just a constant generated by the ```protoc-gen-gogo``` plugin when generating the ```*.pb.go``` files. The constant's definition is in ```proto``` package. It gives us a compile time check between the generated code and the current ```proto``` package.

**Comment by krhubert** [comment]
@jmarais I regenerated proto definitions plus replace the test to make use of reflection.

**Pull request merged**